spi/imx6ull-qspi/qspi.c

```diff
@@ -0,0 +1,326 @@
+
+// TODO ustawić ile ohmów w rejestrze na flashu.
+// TODO find clock frequency - ~50Mhz
+
+static void set_mux(qspi_dev_t dev_no)
+{
+	platformctl_t ctl;
+	int i;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iomux;
+
+	for (i = 0; i < 8; i++) {
+		ctl.iomux.mux = qspi_pctl_mux[dev_no][i].mux;
+		ctl.iomux.mode = qspi_pctl_mux[dev_no][i].val;
+		ctl.iomux.sion = qspi_pctl_mux[dev_no][i].sion;
+		platformctl(&ctl);
+	}
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iopad;
+
+	for (i = 0; i < 8; i++) {
+		ctl.iopad.pad = qspi_pctl_pad[dev_no][i].pad;
+		ctl.iopad.speed = qspi_pctl_pad[dev_no][i].speed;
+		ctl.iopad.dse = qspi_pctl_pad[dev_no][i].dse;
+		ctl.iopad.sre = qspi_pctl_pad[dev_no][i].sre;
+		platformctl(&ctl);
+	}
+}
+
+static void set_clk()
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_devclock;
+	ctl.devclock.dev = pctl_clk_qspi;
+	ctl.devclock.state = 0x03;
+
+	platformctl(&ctl);
+}
+
+int qspi_setLutSeq(const lut_seq_t *lut, unsigned int lut_seq)
+{
+	if (lut_seq >= NUM_LUT_SEQ) {
+		return -1;
+	}
+	if (*QSPI_LCKCR_ADDR & QSPI_LCKCR_LOCK) {
+		*QSPI_LUT_KEY_ADDR = QSPI_LUT_KEY;
+		*QSPI_LCKCR_ADDR = (*QSPI_LCKCR_ADDR & ~(QSPI_LCKCR_UNLOCK | QSPI_LCKCR_LOCK)) | QSPI_LCKCR_UNLOCK;
+	}
+
+	for (int i = 0; i < 4; i++) {
+		*QSPI_LUTn_ADDR(QSPI_SEQ_START_REGN(lut_seq) + i) = (((uint32_t)lut->instrs[2 * i + 1]) << 16) | lut->instrs[2 * i];
+	}
+
+	*QSPI_LUT_KEY_ADDR = QSPI_LUT_KEY;
+	*QSPI_LCKCR_ADDR = (*QSPI_LCKCR_ADDR & ~(QSPI_LCKCR_UNLOCK | QSPI_LCKCR_LOCK)) | QSPI_LCKCR_LOCK;
+}
+
+static void set_IPCR(int seq_num, size_t idatsz)
+{
+	*QSPI_IPCR_ADDR = QSPI_IPCR_SET_IDATSZ(QSPI_IPCR_SET_SEQID(*QSPI_IPCR_ADDR, seq_num), idatsz & 0xFFFF);
+}
+
+void qspi_setTCSH(uint8_t cycles)
+{
+	*QSPI_FLSHCR_ADDR = (*QSPI_FLSHCR_ADDR & ~(0x0F << 8)) | (cycles << 8);
+}
+
+void qspi_setTCSS(uint8_t cycles)
+{
+	*QSPI_FLSHCR_ADDR = (*QSPI_FLSHCR_ADDR & ~(0x0F)) | cycles;
+}
+
+int _qspi_readBusy(qspi_dev_t dev, unsigned int lut_seq, uint32_t addr, void *buf, size_t size)
+{
+	uint8_t to_read, byte, i;
+	uint16_t len = 0;
+	uint32_t reg;
+	int err;
+
+
+	if (dev != qspi_flash_a && dev != qspi_flash_b) {
+		return -ENODEV;
+	}
+	if (lut_seq >= NUM_LUT_SEQ || size > MAX_READ_LEN) {
+		return -1;
+	}
+
+	*QSPI_RBCT_ADDR = (*QSPI_RBCT_ADDR & ~(0x1F) | WATERMARK);
+
+	*QSPI_SFAR_ADDR = qspi_common.flash_base_addr[dev] + addr;
+
+	*QSPI_MCR_ADDR |= QSPI_MCR_CLR_RXF;
+
+	set_IPCR(lut_seq, size);
+
+	while (len < size) {
+		do {
+			reg = *QSPI_SR_ADDR;
+		} while ((reg & QSPI_SR_RXWE) == 0 && reg & QSPI_SR_BUSY);
+
+		for (i = 0; i < WATERMARK + 1 && len < size; i++) {
+			reg = *QSPI_RBDRn_ADRR(i);
+			for (byte = 0; byte < 4 && len + byte < size; byte++) {
+				((uint8_t *)buf)[len + byte] = reg & 0xFF;
+				reg >>= 8;
+			}
+			len += byte;
+		}
+		*QSPI_FR_ADDR |= QSPI_FR_RBDF; /* Buffer pop. */
+	}
+
+	while (*QSPI_SR_ADDR & QSPI_SR_BUSY)
+		;
```

Comment:
MISRA (there should be also some timeout mechanism to be fully MISRA compliant)
```suggestion
	while (*QSPI_SR_ADDR & QSPI_SR_BUSY) {
	}
```